packages/docs/docs/contracts-api/unirep-sol.md

```diff
@@ -103,6 +103,52 @@ function attesterEpochRemainingTime(
 ) public view returns (uint)
 ```
 
+## epochKeyVerifier
+
+A contract address for an epoch key proof verifier. See [IVerifier](/docs/contracts-api/iverifier-sol) for more info.
+
+```sol
+IVerifier public immutable epochKeyVerifier;
+```
+
+Example use:
+
+```sol
+bool valid = unirep.epochKeyVerifier.verifyProof(publicSignals, proof);
```

Comment:
I think we should also let users verify proofs through contract
either
1. create a function in the `Unirep.sol` to let user call verifier: https://github.com/Unirep/Unirep/blob/7945d2e7d95981e489af0d231f8734a11db94a50/packages/contracts/contracts/Unirep.sol#L439
2. export `IVerifier` abi so that user can reconstruct `const verifier = new Contract(address, abi)`

Reply:
I disagree, the user would be better off calling the verifier manually. If we implement verifier functions in the unirep contract it should verify the proof AND verify the existence of the state root for the attester+epoch.

We don't need to check `isSNARKField` ourselves, the verifier contract checks this already.

Reply:
yeah so if they want to call the verifier they should know the verifier abi
they should also check the state root for the attester+epoch by theirselves
it depends on their applications

Reply:
the `verifyEpochKeyProof ` is to make sure the user uses the same verification key as `Unirep.sol`
and it performs like `Prover.verifyProof()`

Reply:
>  they should also check the state root for the attester+epoch by theirselves

I think this is unreasonable. If we provide a function `verifyEpochKey` in the unirep contract, it's extremely misleading if it doesn't verify everything.

> the verifyEpochKeyProof is to make sure the user uses the same verification key as Unirep.sol

If an external contract does `epochKeyVerifier.verifyProof()` they're using the same verification key right?

Reply:
> I think this is unreasonable. If we provide a function verifyEpochKeyProof in the unirep contract, it's extremely misleading if it doesn't verify everything.

It provides `Prover.verifyProof()` without user having `snarkjs` and `vkey`
On second thought, I agree that `verifyEpochKeyProof ` can provide state root check 

> If an external contract does epochKeyVerifier.verifyProof() they're using the same verification key right?

yes I think so